src/app/avatar/page.tsx

```diff
@@ -1,78 +1,140 @@
+  const [connectionState, setConnectionState] = useState("");
+
+  useEffect(() => {
+    // Cleanup peer connection on component unmount
+    return () => {
+      clientRef.current?.close();
+    };
+  }, []);
 
-  function createEnumButtons<T extends string | number>(enumType: Record<string, T>): JSX.Element[] {
+  function createEnumButtons<T extends string | number>(
+    enumType: Record<string, T>,
+  ): JSX.Element[] {
     return Object.values(enumType).map((key) => (
-      <button 
+      <button
         key={key as string}
-        onClick={() => setSelectedProvider(key as Provider)}
-        className={`px-4 py-2 rounded-md mr-2 
-                   ${selectedProvider === key ? 'bg-fixie-fresh-salmon text-white'
-                                             : 'bg-gray-200 hover:bg-gray-300'
-                   }`}  
+        onClick={() => generate(key as Provider)}
+        className="mr-1 rounded-md bg-fixie-fresh-salmon hover:bg-fixie-ripe-salmon px-3 py-2 text-sm font-semibold text-white shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-fixie-fresh-salmon"
       >
-        {key as string} 
+        {key as string}
       </button>
     ));
   }
+  const createConnectionText = () => {
+    return currentProvider ? `${currentProvider}: ${connectionState}` : "";
+  };
+  const createClient = (provider: Provider) => {
+    switch (provider) {
+      case Provider.DIDTalks:
+        return new DidTalksClient(mediaElementRef.current!, DID_SOURCE_URL);
+      case Provider.DIDClips1:
+        return new DidClipsClient(
+          mediaElementRef.current!,
+          DID_PRESENTER_ID1,
+          DID_DRIVER_ID1,
+        );
+      case Provider.DIDClips2:
+        return new DidClipsClient(
+          mediaElementRef.current!,
+          DID_PRESENTER_ID2,
+          DID_DRIVER_ID2,
+          DID_VOICE_ID2,
+        );
+      case Provider.HeyGen:
+        return new HeyGenClient(mediaElementRef.current!);
+      case Provider.Microsoft:
+        return new AzureClient(mediaElementRef.current!);
+      default:
+        throw new Error("Not implemented yet");
+    }
+  };
+  const connect = async (provider: Provider) => {
+    clientRef.current = createClient(provider);
+    setCurrentProvider(provider);
+    setConnectionState("connecting");
+    clientRef.current.addEventListener(
+      "connectionState",
+      (evt: CustomEventInit<RTCPeerConnectionState>) =>
+        setConnectionState(evt.detail!),
+    );
+    await clientRef.current.connect();
+    while (!clientRef.current.connected) {
+      await new Promise((resolve) => setTimeout(resolve, 100));
+    }
+  };
+  const generate = async (provider: Provider) => {
+    if (!clientRef.current || provider != currentProvider) {
+      clientRef.current?.close();
+      await connect(provider);
+    }
+    await clientRef.current!.generate({ text: { text } });
+  };
+
```

Comment:
Should we add a method to teardown the client connection once we switch to a different provider?

Reply:
I think that's on line 105, right?